src/backends/gpu/kernels/reduce_by_key.hpp

```diff
@@ -0,0 +1,132 @@
+#pragma once
+
+#include <cstdint>
+#include "detail.hpp"
+
+namespace nest {
+namespace mc {
+namespace gpu {
+
+namespace impl{
+
+// return the power of 2 that is less than or equal to i
+__device__ __inline__
+unsigned rounddown_power_of_2(unsigned i) {
+    // handle power of 2 and zero
+    if(__popc(i)<2) return i;
+
+    return 1u<<(31u - __clz(i));
```

Comment:
Sure, I have made the change.

The API accepting only `int` is an oversight. It behaves as "expected", as the following code illustrates:
```
#include <limits>
#include <cstdio>
#include <iostream>

__global__
void krnl(unsigned x) {
    printf("%-12d : %d\n", x, __clz(x));
}

int main() {
    unsigned r=1;
    krnl<<<1,1>>>(0);
    for (int i=0; i<32; ++i) {
        krnl<<<1,1>>>(r);                                                                                
        r <<= 1;
    }   

    krnl<<<1,1>>>(std::numeric_limits<unsigned>::max());
    cudaDeviceSynchronize();

    return 0;
}

=== output ===
> nvcc clz.cu -arch=sm_35 && ./a.out
0            : 32
1            : 31
2            : 30
4            : 29
8            : 28
16           : 27
32           : 26
64           : 25
128          : 24
256          : 23
512          : 22
1024         : 21
2048         : 20
4096         : 19
8192         : 18
16384        : 17
32768        : 16
65536        : 15
131072       : 14
262144       : 13
524288       : 12
1048576      : 11
2097152      : 10
4194304      : 9
8388608      : 8
16777216     : 7
33554432     : 6
67108864     : 5
134217728    : 4
268435456    : 3
536870912    : 2
1073741824   : 1
-2147483648  : 0
-1           : 0
```